Add doc for docker-compose, makefile commands, and images in README.md (#305)
apache/apisix-docker: README.md

## What is APISIX

Apache APISIX is a dynamic, real-time, high-performance API gateway. APISIX provides rich traffic management features such as load balancing, dynamic upstream, canary release, circuit breaking, authentication, observability, and more.

See [the APISIX website](https://apisix.apache.org/) for more info.

## Image variants

The APISIX image comes in many flavors, each designed for a specific use case.

`apisix:<version>`

> This is the defacto image. If you are unsure about what your needs are, you probably want to use this one. It is designed to be used both as a throw away container (mount your source code and start the container to start your app), as well as the base to build other images off of.

Reviewer: ```suggestion
This is the default image. If you are unsure about what your needs are, this is your go-to option.you can use it as a throw away container (mount your source code and start the container to start your applications), as well as the base to build other images of.
```

`apisix:<version>-alpine`

> This image is based on the popular Alpine Linux project, available in the alpine official image. Alpine Linux is much smaller than most distribution base images (~5MB), and thus leads to much slimmer images in general.

Reviewer: ```suggestion
This image is based on the popular Alpine Linux project. Since Alpine Linux is much smaller than most distribution Linux images (~5MB), you can build smaller images with it.
```

> This variant is useful when final image size being as small as possible is your primary concern. The main caveat to note is that it does use musl libc instead of glibc and friends, so software will often run into issues depending on the depth of their libc requirements/assumptions. See this Hacker News comment thread for more discussion of the issues that might arise and some pro/con comparisons of using Alpine-based images.

Reviewer: ```suggestion
This variant is useful when storage compatibility is your primary concern. The main caveat to note is that it does use `musl libc` instead of `glibc` and friends, so software will often run into issues depending on the depth of their `libc requirements/assumptions`. See this Hacker News comment thread for more discussion of the issues that might arise and some advantages and disadvantagess comparisons of using Alpine-based images.
```

Reviewer: would be nice to have a "Hacker News" hyperlink here.

> To minimize image size, it's uncommon for additional related tools (such as git or bash) to be included in Alpine-based images. Using this image as a base, add the things you need in your own Dockerfile (see the alpine image description for examples of how to install packages if you are unfamiliar).

Reviewer: ```suggestion
To minimize the image size, additional tools, such as git and bash, are not included in Alpine-based images. Using this image as a base, add the things you need in your own Dockerfile (see the alpine image description).
```

## How to run APISIX

APISIX can be run using docker compose or using the `all-in-one` image. It is recommended to use docker compose to run APISIX, as `all-in-one` deploys all dependencies in a single container and should be used for quick testing.
If you want to manually deploy services, please refer to [this guide](https://github.com/apache/apisix-docker/blob/master/docs/en/latest/manual.md).

### Run APISIX with docker-compose

[The apisix-docker repo](https://github.com/apache/apisix-docker/blob/master/example) contains an example docker-compose file and config files that show how to start APISIX using docker compose. For the sake of completeness, this docker-compose file also starts [APISIX dashboard](https://hub.docker.com/r/apache/apisix-dashboard), which is a frontend interface that makes it easy for users to interact with APISIX, along with Prometheus and Grafana.

To try out this example:

> 1. Clone the [repo](https://github.com/apache/apisix-docker) and cd into the root folder.
>   

Reviewer: ```suggestion
1. Clone the [repo](https://github.com/apache/apisix-docker) and cd into the root folder.
    ```
    git clone 'https://github.com/apache/apisix-docker'
    cd apisix-docker
    ```
```

2. Start APISIX.
    ```
    cd example

    docker-compose -p docker-apisix up -d
    ```

3. Check if APISIX is running properly by running this command and checking the response.
    ```
    curl "http://127.0.0.1:9080/apisix/admin/services/" -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1'
    ```
     The response indicates that apisix is running successfully:
    ```
    {
      "count":0,
      "action":"get",
      "node":{
        "key":"/apisix/services",
        "nodes":[],
        "dir":true
      }
    }
    ```

The [example docker compose file](https://github.com/apache/apisix-docker/blob/master/example/docker-compose.yml) defines several services: `apisix-dashboard, apisix, etcd, web1, web2, prometheus, and grafana`:

> - `apisix-dashboard, apisix, etcd` are the essential services required for starting apisix-dashboard, apisix, etcd.

Reviewer: ```suggestion
- `apisix-dashboard, apisix, etcd` are the essential services required for starting apisix-dashboard, apisix, and etcd.
```

If you want to use a config file located at a different path, you need to modify the local config file path in the `volumes` entry under the corresponding service.

### Run APISIX with all-in-one command

A quick way to get APISIX running on alpine is to use the `all-in-one` docker image, which deploys all dependencies in one Docker container. You can find the dockerfile [here](https://github.com/apache/apisix-docker/blob/master/all-in-one/apisix/Dockerfile). The image utilizes [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/), building APISIX layer and etcd layer first, then copying the nesessary artifacts to the alpine layer.

To try out this example:

> 1. Clone the [apisix-docker repo](https://github.com/apache/apisix-docker/blob/master) and cd into the root folder.

Reviewer: ditto

Author: Okay. I changed this to  "1. Make sure that you are in the root folder of apisix-docker".

2. `make build-all-in-one` to build the `all-in-one` image.

3. Launch the APISIX container:

    ```sh
    docker run -d \
    -p 9080:9080 -p 9091:9091 -p 2379:2379 \
    -v `pwd`/all-in-one/apisix/config.yaml:/usr/local/apisix/conf/config.yaml \
    apache/apisix:whole
    ```

4. Check if APISIX is running properly by running this command and checking the response.
    ```
    curl "http://127.0.0.1:9080/apisix/admin/services/" -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1'
    ```
     The response indicates that apisix is running successfully:
    ```
    {
      "count":0,
      "action":"get",
      "node":{
        "key":"/apisix/services",
        "nodes":[],
        "dir":true
      }

>     }

Reviewer: ```suggestion
    }
"```"  
```

Reviewer: add ``` after line 108 to surround the end of code block.

## Summary
This repo includes:

> 1. docker images for APISIX and APISIX dashboard (on alpine and centos). 

Reviewer 2: No indent at the end

Author: I don't think indentation is required for an ordered list: https://www.markdownguide.org/basic-syntax/#ordered-lists
It's indented automatically.

2. examples that show how to get APISIX and APISIX dashboard up and running. 
3. a list of commands that allow users to easily build, save, and tar the docker images.

## How To Build Images

> **Note**: Docker images are not official ASF releases but provided for convenience. Recommended usage is always to build the source. \

Reviewer 2: Why are there two Notes?
And a `\` at the end?

Author: They were originally in different places. I moved them to the same place because I think they are both related to images/building images. I agree that it's confusing to have two Notes side by side. I will move them elsewhere.

-v `pwd`/all-in-one/apisix-dashboard/conf.yaml:/usr/local/apisix-dashboard/conf/conf.yaml \
apache/apisix-dashboard:whole
```
## Useful commands

Below are some useful commands which build, push, and tar your updated images.
As an example, these are the commands for apisix-centos images:

-   ```make build-on-centos``` : Build apache/apisix:xx-centos image. 

-   ```make push-on-centos```: Build and push apache/apisix:xx-centos image.

> -  ```make save-centos-tar```:  Save apache/apisix:xx-centos image to a tar archive located at ```./package``` . 

Reviewer 2: Some questions here?
Please note that the indentation remains consistent

Author: Thanks! Will double check my indentations
